### ISSUE TITLE
Use epmfs, when no free space fall back to mfs

### ISSUE THREAD

**Issue by cyberius0** [trapexit/mergerfs#634]
Hello,

I would like to use the policy "epmf" (existing path, most free space).

The problem is, when the drive containing this path gets full or reaches "minfreespace", I can't write anything anymore into this path, I get the error "No space left on device" (in Samba and in Terminal).

The manual says, that in this case (no more free space left) mergerfs should fall back to "mfs", where it chooses a free disk, that doesn't contain that path yet.

But it doesn't happen here. What am I doing wrong?

To sum it up: 
I want to use "epmf". When the disk containing the path gets full, it should switch to policy "mfs".

How can I do this?

Thank you

**Comment by trapexit** [comment]
What's the purpose for doing that. In the log run that means it's the same as not using epmfs. If you have other things, like backups, relying on path preservation then you risk messing that up because you won't know when a new drive is used for that path.

**Comment by trapexit** [comment]
I'm not opposed to adding this I just want to make sure it's understood what the long term consequences would be. It's mostly a difference of degrees in where data ends up rather than kind. It will limit the entropy a bit but not strictly manage it like it currently does.

**Comment by cyberius0** [comment]
I currently use "category.create=mfs". I thought it would be nice if mergerfs uses the existing path as long as free space is available but if there isn't enough space I don't want to be bothered with the no free space errors. So something like "re-use existing paths if possible, fall back to other disks if necessary". This way I know I have to resort all files from time to time, but much less work as with mfs, where all files always get placed on another disk.

**Comment by trapexit** [comment]
Like I mentioned I'm not opposed to it but the reason it defaults to epmfs is because that's how I started. That's what I thought I wanted and then as my hoard grew I realized the entropy is going to happen and the only real benefit it seems of colocation of files on drives is if you have no backup and wish to lose all of X instead of 1/Nth. Usually when people partition data into important and not so important they do so at the directory level rather than drive level.

Anyway... its probably not that much work and it shouldn't really add any real cost to the policy. I'll look into it for a future release.

And thanks. Not sure about masterpiece but I'm glad more than just myself get a use out of it.

**Comment by spcqike** [comment]
Hello

I just want to say that’s exactly what I am looking for, too. 
I use a baremetal openmediavault Server for my backup. 
My main NAS is running a raid5, so i want to keep costs low and use old disks for my backup. So in the worst case (failed raid5 restore and failed backup) for media files I’d rather like to loose a complete season/series instead of one episode in every season. 

It’s easier to rip a Serie again, than ripping single files from all series  :)

**Comment by trapexit** [comment]
This wouldn't behave that way. As I mentioned before... mfs will create the path on any drive with the most free space. If you're writing files to a single path then you'll now have 1 season or series spread across 2 drives. Once that fills... 3 drives. Etc.

Someone asked for a consolidation script but the difference would be one of degrees, not kind.

**Comment by spcqike** [comment]
ok, so either i misunderstood the last posts, or you misunderstood me :)
i thought the general idea was epmfs with mfs as fallback.

so, in the first step it tries to keep everything within the directory / disk (epmfs). when the disk runs out of space, it creates the needed path on a second disk(mfs) and stays there(epmfs, again). 

isn't that, what cyberious0 was taking about?

**Comment by trapexit** [comment]
That's exactly what we're talking about but I think you misunderstand what it will result in.

Path preservation works on all paths found. If mfs is used a new path is created on a new branch.

Assume branches A, B, & C, path is on A only. A fills up. New file created and mfs is used due to that. B or C will be chosen based on available space. Now A & B have the path. Files will go to B now. Now you have files on both A and B. B fills... now on A, B, & C.

So... it limits the spread but it does not fundamentally differ from mfs. Which should be obvious if we're saying that mfs is the fallback to epmfs. In the end it makes epmfs = mfs. It just changes the layout slightly and when they end up in a location.

**Comment by spcqike** [comment]
but it would keep files together, in the first step. right?
just using mfs would split everything over all branches. from the very beginning.

right now, i use "least free storage" so that seasons within a series a kept together (initial backup). when i remove and add stuff on my main NAS and start the backup-job, it would just use the disk with "least free storage" and split seasons apart. 
i thought, the "epmfs // mfs" constellation would recognize, that a single file belongs to an existing path/series/season and put it there (if there is free space)

**Comment by trapexit** [comment]
If it's created a new path on a new drive the files would split across those paths based on the policy. In this case mfs. If you create a file in the season directory and epmfs fails then mfs will create that same path on another drive. Now there are 2 of the same relative path. epmfs will now choose the one of those two with the mfs. You will have two files in the same relative path on two separate drives.

Just play out the situation given the policy descriptions.

Have you considered using mkdir=mfs and create=epmfs? 

No matter what... If you have path preservation fall back to non path preservation you are, in the long run, behaving the same as non path preservation. It's a matter of degrees different, not kind.

**Comment by trapexit** [comment]
mergerfs can't predict the future. If you have a directory and start adding files. You will add 10 1GB files. You have only 5GB till minfreespace is triggered (and nothing else is writing to the drive) then once you write 5 files the 6th (with mfs or epmfs w/ fallback) will result in drive 2 having the same relative path and the next 5 files will be written on drive 2. Now you have 5 on A and 5 on B. There was no way for mergerfs to manage this situation otherwise.

**Comment by fh-fh** [comment]
I also use the nice mergerfs (Thank you for making this.) with snapraid and miss the fallback which was removed in 2.25.0
One advantage of snapraid is that you could hold the fragmentation of data between the disks low.
Of course in case of fallback you have then files in two branches but this is in my eyes better than an out of space error and no file is created.
Do you see any chance to implement the fallback again?

**Comment by trapexit** [comment]
The fallback was not consistent with the intended behavior and was a bug. It'd be like copying a file to an entirely different filesystem because you run out of space. The point of path preservation is to have specific control. Fallback ultimately negates that. If you don't care where files end up just turn off path preservation. If you have proper backup or are doing really niche things there isn't a reason to use path preservation. If you are doing niche things or want to limit where data goes why would you want to silently increase the pool size of a path? If you ever delete data or move it it will then start spreading files across the two drives, then 3, then 4, etc. That's exactly why I stopped using path preservation. It becomes increasingly pointless.

I don't understand what you mean by "hold the fragmentation of data between the disks low". On average that's only true if you have 1 path per drive or maybe use policy ff. Otherwise files will naturally end up across the drives. Why would fragmentation of random files across drives matter to snapraid?

If I did add it it'd be a new policy that explicitly says it's for sortof path preservation and clearly document the downsides.

**Comment by trapexit** [comment]
What specifically are people trying to accomplish? I'm still not understanding the core reasons for wanting this behavior. If you're just looking to mostly colocate files in a single directory on a single drive why not use `mfs` for `mkdir` and `epmfs` for `create`? Won't work well if you put tons of files into a single directory though. Then again you could just create separate pools for different behaviors.

I suppose creating a policy like "highest link count" applied to create may behave similar without the path preserving limitation. But again... where does it stop? You get all files for season 1 on drive A but season 2 on drive B? "But I want to keep all episodes of the show one drive but don't want to have all files in one directory."

This stuff starts getting very impractical and bespoke. Unless I added per directory policies which wouldn't be used by most people and could be generally replicated via multiple mounts.

**Comment by zofrex** [comment]
When I first set up mergerfs, the README described epmfs working in this way, and I remember thinking I wanted it then. The more I read what you've written about it, the less certain I am that I want it, but given that I have a use-case for it that I don't believe I've seen articulated yet, I think perhaps it's worth outlining:

I want to minimise the work needed to restore data in the case of drive failure, but balance that with not having to do extra manual work when I add new data.

Why does epmfs minimise work restoring data for me? Because replacing 1 season of 10 episodes of a TV show is significantly less work than replacing 1 episode each from 10 different seasons. I think this might apply to a few people, depending what they are storing, but that might involve having to load 1 DVD and rip it vs loading 10 DVDs, or grabbing 1 torrent vs 10.

Why is strict epmfs not sufficient for me as is? Because being able to treat my drives as one giant pool of data and not worry about what is on which drive is one of the greatest benefits of mergerfs, and although sometimes I have to deal with the reality that they are different drives, and files are on particular drives, I like to avoid that as much as possible. In particular being able to blindly write to the mergerfs mount and not think about the underlying drives is very nice.

What I _think_ will happen if epmfs with fallback to mfs existed and I used it: most of the time, all of a given TV show will be on a single drive. Nearly all of the time, a given TV show season will be on a single drive. Sometimes, a particular season will be split across two drives. More often, a TV show will be split across two drives, or sometimes more than two.

This sounds desirable to me. Although it won't be perfect, in the event of a data loss, I'd mostly be dealing with complete seasons, or halves of seasons. If I used mfs on its own, I believe I'd probably have to replace one or two episodes from almost every season.

For additional context, I (generally) only add data and very rarely delete anything. I also try to keep a decent amount of free space in the overall pool, and add new drives incrementally as the pool fills up. I think this way of using mergerfs increases the chances that seasons will be mostly kept together if "epmfs with fallback to mfs" existed, but I'm not 100% sure about that. You have certainly thought about this a lot more than me, so if you don't think epmfs with fallback will help my scenario then I believe you, likewise if you say that mfs for mkdir and epmfs for create is sufficient. I thought it might be helpful to have a specific user story complete with context and motivation, though.

**Comment by trapexit** [comment]
I'm completely familiar with the usecase but there are multiple existing ways to get that without the downsides that randomly ignoring the paths.

Some people use ff or lfs. You'll get split directories when the drive fills. If you're not deleting or moving stuff then that happens once in a blue moon. If you do delete stuff it's not a good idea because you'll more regularly split dirs every time a drive goes below minfreespace.

Some people change the mkdir policy. Some use rand. How well that works depends on how you lay out files. If you mostly keep related files in a directory like TV seasons, music albums, movies, etc. then it works fine. If you just shove all the movies in one directory like I do that wouldn't work as well without seeding the movie path because otherwise it'd just put the movies into one drive.

Having a blanket non-path preserving policy fallback means different things depending on how it gets triggered. And you don't have delete things for it to end up erasing the path preservation. If you create file /a/b/c/d/e/f/g/h and /a/b/c/d/e/f/g doesn't exist on epmfs drive then /a/b/c/d/e/f/g is created on mfs. Each of those subdirs are now possible locations for any new dir or file. Every path on that now excluded drive is possibly going to be copied to another drive now. If you delete anything you'll now have 2 drives as possible locations for both files and dirs. Over time you will end up having a lot more files spread around than you expect. This happened to many people, including myself. The original fallback behavior wasn't accurate to the usecase it intended to solve and how it should have been described. All the people who used it for that purpose ended up complaining and/or realized it didn't really give them what they thought and most moved to lus or mfs. 

People seem to forget that mkdir has polices and ultimately if you want to really keep files within a single directory in a single drive then you don't change the create policy. You change the mkdir policy.

If people really don't care about the directory sprawl that will occur over time then as I've said in other threads I'll add fallbacks in some way but I don't believe most people really understand what that means. And those above are different from the tree walkback others have discussed which would recreate the directory path only if the path preservation failed due to hitting minfreespace. But even that is just a difference of degrees from current path preservation. You'll still end up with dir spread and files spread around and you will have to seed dirs when the walk back depth is hit and those drives are full. It just won't be as often as now. It'd lead to less spread than fallback but doesn't give you what you want necessarily.

**Comment by trapexit** [comment]
Should also be mentioned that even if you don't have backup many people use something like Sonarr, Radarr, etc. that can download bespoke missing content it minimizes the issue of losing disperse data as they are designed to deal with precisely that kind of situation.

**Comment by Treverr** [comment]
I have this use case as well where I have 2 drives similary structured (i.e. Media/TV Shows and Media 2/TV Shows) and I want to do the same thing. Media is almost full so if I created a mount point for this to combine it all, what happens when its full and it tries to put a new episode into a path already on media? I am using Sonarr to manage the shows. My concern is that continuing tv shows will fill the drive up and then need to flip to a second drive which Sonarr doesn't handle at this point.

If I set `mfs` for `mkdir` and `epmfs` for `create` will this achieve what I'm trying to do here?
